app/human-detector-app/App.tsx

```diff
@@ -2,30 +2,75 @@ import * as React from 'react';
 import { View, Text, StyleSheet, Platform } from 'react-native';
 import { NavigationContainer } from '@react-navigation/native';
 import { createNativeStackNavigator } from '@react-navigation/native-stack';
+import { TokenResponse } from 'expo-auth-session';
+import JWT from 'expo-jwt';
+import * as CryptoJS from 'crypto-js';
+import User from './classes/User';
 import CameraScreen from './screens/CameraScreen';
 import GroupScreen from './screens/GroupScreen';
 import LoginScreen from './screens/LoginScreen';
+import { getExponentPushToken, sendExpoNotifToken } from './src/notifications/notifTokenInit';
 
 const Stack = createNativeStackNavigator();
 
 export default function App(): React.ReactElement {
+  let isUserSignedIn = false;
+
+  // Update upon getting a token
+  const [tokenResponse, setTokenResponse] = React.useState(null);
+
+  // Check if the token is expired
+  if (tokenResponse != null) {
+    // TODO: refresh token if expired
+
+    // if not expired, then user is signed in
+    isUserSignedIn = true;
+  }
+
+  // If the user isn't logged in
+  if (!isUserSignedIn) {
+    // Open the login screen
+    return (
+      <NavigationContainer>
+        <Stack.Navigator>
+          <Stack.Screen
+            name="Login"
+            options={{
+              title: 'EyeSpy',
+              headerStyle: {
+                backgroundColor: '#1E90FF',
+              },
+              headerTintColor: '#fff',
+              headerTitleStyle: {
+                fontWeight: 'bold',
+              },
+            }}
+          >
+            {(props) => <LoginScreen setTokenResponse={setTokenResponse} />}
+          </Stack.Screen>
+        </Stack.Navigator>
+      </NavigationContainer>
+    );
+  }
+
+  // If logged in, make the user and put them into the app
+
+  // Decode Base64
+  const words = CryptoJS.enc.Base64.parse(tokenResponse.idToken.split('.')[1]); // Get the payload
+  const textString = CryptoJS.enc.Utf8.stringify(words);
+
+  const user = new User(
+    JSON.parse(textString).preferred_username,
+    JSON.parse(textString).sub,
+    isUserSignedIn
+  );
+
+  sendExpoNotifToken(user.userID).catch(() => console.error('Error in sending expo token!'));
```

Comment:
We should put this call in a `useEffect` hook (probably depending on `isUserSignedIn`, or the `User` object if you apply those suggestions). Idk about you, but I have a hard time keeping track of when re-renders will happen in this component. It would be easier to reason about when this function is called if we threw it in a `useEffect` hook and had a dependency array. We might accidentally spam our API if a bunch of re-renders happen in this component as it is right now

Reply:
Oh yeah that makes sense.  Maybe I could send the token in the useEffect in the KeyCloakButton component as seen [https://github.com/tucker-moore/human-detector/blob/dfe77bc307237c52ac1642ba2a16117e20ecf3a8/app/human-detector-app/components/KeyCloakButton.tsx#L50](here)

Reply:
You *could*, but KeyCloakButton's job should probably just be to get the token set

Reply:
Ok yeah you're right I'll put this in a useEffect: 936f6631b01a93fc43d6c096c0db4c8697bc7bee